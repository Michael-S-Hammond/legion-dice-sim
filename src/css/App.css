.graph-comparison-button-add {
  background-image: url("../img/clipboard-plus.svg");
  display: flex;
  width: 40px;
  height: 40px;
  background-repeat: no-repeat;
  background-size: 85%;
  background-position-x: 2px;
  background-position-y: 2px; }

.graph-comparison-button-compare {
  background-image: url("../img/clipboard-data.svg");
  display: flex;
  width: 40px;
  height: 40px;
  background-repeat: no-repeat;
  background-size: 85%;
  background-position-x: 2px;
  background-position-y: 2px; }

.roll-button {
  display: flex;
  height: 40px;
  justify-content: center;
  align-items: center;
  font-weight: bold;
  font-size: large; }

.drop-down-label {
  font-size: large; }

select {
  height: 30px;
  margin: auto 0; }

.popover-button {
  display: flex;
  height: 28px;
  width: 28px;
  background-repeat: no-repeat;
  background-size: 85%;
  background-position-x: 2px;
  background-position-y: 2px;
  background-image: url("../img/question-circle.svg"); }
