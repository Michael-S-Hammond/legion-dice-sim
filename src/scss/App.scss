.roll-button {
  display: flex;
  height: 40px;
  justify-content: center;
  align-items: center;
  font-weight: bold;
  font-size: large;
}

.export-button {
  display: flex;
  width: 30px;
  height: 30px;
  background-repeat: no-repeat;
  background-size: 85%;
  background-position-x: 2px;
  background-position-y: 2px;
  background-image: url('../img/download.svg');
}

.drop-down-label {
  font-size: large;
}

select {
  height: 30px;
  margin: auto 0;
}
