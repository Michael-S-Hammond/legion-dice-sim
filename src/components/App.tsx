import '../css/App.css';

import * as MJS from "mathjs";
import React from 'react';

import * as DU from '../code/DiceRoller';
import * as T from '../code/Types';
import * as AS from './AppStateManager';

import Attack from './Attack';
import Combat from './Combat';
import Defense from './Defense';
import DiceResults from './DiceResults';
// import Notification from './Notification';
import Header from './Header';

import { Telemetry } from '../tools/Telemetry';

class App extends React.Component<any, AS.AppState> { // eslint-disable-line @typescript-eslint/no-explicit-any
  private _stateManager: AS.AppStateManager;
  private _diceResultsRef: React.RefObject<DiceResults>;
  private _telemetry: Telemetry;

  constructor(props: any) { // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
    super(props);

    this._stateManager = new AS.AppStateManager((newState: AS.AppState) => this.setState(newState));
    this.state = this._stateManager.state;

    this._diceResultsRef = React.createRef();
    this._telemetry = new Telemetry();
  }

  private handleShowExpectedRangeChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    const show = event.target.checked;
    this._stateManager.behaviorEventHandlers.handleShowExpectedRangeChanged(show);
    this._diceResultsRef.current?.forceChartUpdate();
  }

  private handleSimplifiedViewChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const show = event.target.checked;
    this._stateManager.behaviorEventHandlers.handleShowSimplifiedViewChange(show);
  }

  private handleDiceCountChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const count = Number(event.target.value);
    this._stateManager.behaviorEventHandlers.handleDiceCountChange(count);
  }

  private rollAttackDice() {
    const diceRoller = new DU.DiceRoller();

    performance.mark('start roll');
    const outputs = diceRoller.simulateAttacks(
      this.state.diceRolls,
      this.state.inputs
    );
    performance.mark('end roll');

    performance.measure('roll duration', 'start roll', 'end roll');
    const rollDuration = MJS.round(performance.getEntriesByName('roll duration', 'measure')[0].duration, 3);
    this._telemetry.trackEvent("RollDice", {
      Rolls: this.state.diceRolls,
      Duration: rollDuration,
      SimpleView: this.state.view.showSimpleView,
    });

    this._diceResultsRef.current?.scrollIntoView();
    this._diceResultsRef.current?.forceChartUpdate();

    this._stateManager.updateOutputs(outputs);
  }

  render() : JSX.Element {
    return (
      <main role="main">
        <Header></Header>
        {/* <Notification message='Rerolls received a substantial update. More coming soon.'></Notification> */}
        <div className="container">
          <div className="row">
            <div className="col-md-8 offset-md-2 mt-3">
              <div key='showSimplifiedViewToggle-abilitytoggle' className="d-flex justify-content-center my-auto custom-control custom-switch">
                <input key='showSimplifiedViewToggle-toggle-input' type="checkbox"
                  className='custom-control-input my-auto'
                  id="showSimplifiedViewToggle"
                  checked={this.state.view.showSimpleView}
                  onChange={this.handleSimplifiedViewChange}></input>
                <label key='showSimplifiedViewToggle-toggle-label'
                  className='custom-control-label drop-down-label mx-2 my-auto'
                  htmlFor="showSimplifiedViewToggle"
                >Simplified View</label>
              </div>
            </div>
          </div>
          <div className="row">
            <div className={`col-md-5 offset-md-1 col-lg-4 ${this.state.view.showSimpleView ? 'offset-lg-2' : 'offset-lg-0'}`}>
              <Attack
                showSimpleView={this.state.view.showSimpleView}
                input={this.state.inputs.offense}
                eventHandlers={this._stateManager.attackEventHandlers}
              ></Attack>
            </div>
            <div className="col-md-5 col-lg-4">
              <Defense
                showSimpleView={this.state.view.showSimpleView}
                input={this.state.inputs.defense}
                eventHandlers={this._stateManager.defenseEventHandlers}
              ></Defense>
            </div>
            <div className={`col-md-6 offset-md-3 col-lg-4 offset-lg-0 ${this.state.view.showSimpleView ? 'collapse' : 'collapse.show'}`}>
              <Combat
                input={this.state.inputs.combat}
                eventHandlers={this._stateManager.combatEventHandlers}
              ></Combat>
            </div>
          </div>
          <div className="d-flex flex-wrap justify-content-center my-3">
            <span className="mx-2 my-auto drop-down-label">Rolls:</span>
            <select value={this.state.diceRolls} className="rounded-lg mr-4 px-2"
              onChange={this.handleDiceCountChange}>
              <option value="1">1</option>
              <option value="10000">10,000</option>
            </select>
            <button className="btn btn-secondary border border-secondary rounded-lg mx-2 px-3 roll-button"
              onClick={() => this.rollAttackDice()}>Roll</button>
            <div key={`showExpectedRangeToggle-abilitytoggle`} className="d-flex justify-content-center my-auto custom-control custom-switch">
              <input key={`showExpectedRangeToggle-toggle-input`} type="checkbox"
                className={`custom-control-input my-auto ${this.state.resultsVisibility === T.ResultOutput.Graph ? 'collapse.show' : 'collapse'}`}
                id="showExpectedRangeToggle"
                checked={this.state.view.showExpectedRange}
                onChange={this.handleShowExpectedRangeChanged}></input>
              <label key={`showExpectedRangeToggle-toggle-label`}
                className={`custom-control-label drop-down-label mx-2 my-auto ${this.state.resultsVisibility === T.ResultOutput.Graph ? 'collapse.show' : 'collapse'}`}
                htmlFor="showExpectedRangeToggle"
              >Show expected range</label>
            </div>
          </div>
          <DiceResults
            ref={this._diceResultsRef}
            results={this.state.outputs}
            visibility={this.state.resultsVisibility}
            showExpectedRange={this.state.view.showExpectedRange}
          ></DiceResults>
        </div>
      </main>
    );
  }
}

export default App;
