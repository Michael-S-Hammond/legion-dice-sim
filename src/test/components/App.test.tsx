import React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import App from 'components/App';

describe('App', () => {
    it('matches the snapshot', () => {
        const app = shallow(<App/>);
        const snapshot = shallowToJson(app);
        expect(snapshot).toMatchSnapshot();
    });

    it('handles simple view', () => {
        const app = shallow(<App/>);
        app.find('#showSimplifiedViewToggle').simulate('change', { target: { checked: true } });

        const snapshot = shallowToJson(app);
        expect(snapshot).toMatchSnapshot();
    });
});
