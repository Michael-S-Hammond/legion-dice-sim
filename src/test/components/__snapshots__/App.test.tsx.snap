// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App handles simple view 1`] = `
<main
  role="main"
>
  <Header />
  <ProfileSelectorDialog
    applyProfile={[Function]}
    id="attackProfileDialog"
    upgradeAllowListName="attack"
  />
  <ProfileSelectorDialog
    applyProfile={[Function]}
    id="defenseProfileDialog"
    upgradeAllowListName="defense"
  />
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-8 offset-md-2 mt-3"
      >
        <div
          className="d-flex justify-content-center my-auto custom-control custom-switch"
          key="showSimplifiedViewToggle-abilitytoggle"
        >
          <input
            checked={true}
            className="custom-control-input my-auto"
            id="showSimplifiedViewToggle"
            key="showSimplifiedViewToggle-toggle-input"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="custom-control-label drop-down-label mx-2 my-auto"
            htmlFor="showSimplifiedViewToggle"
            key="showSimplifiedViewToggle-toggle-label"
          >
            Simplified View
          </label>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-5 offset-md-1 col-lg-4 offset-lg-2"
      >
        <Attack
          eventHandlers={
            Object {
              "handleBlastChange": [Function],
              "handleCriticalXChange": [Function],
              "handleCriticalXValueChange": [Function],
              "handleDuelistChange": [Function],
              "handleHighVelocityChange": [Function],
              "handleImpactXChange": [Function],
              "handleImpactXValueChange": [Function],
              "handleIonXChange": [Function],
              "handleIonXValueChange": [Function],
              "handleJediHunterChange": [Function],
              "handleLethalXChange": [Function],
              "handleLethalXValueChange": [Function],
              "handleMakashiMasteryChange": [Function],
              "handleMarksmanChange": [Function],
              "handlePierceXChange": [Function],
              "handlePierceXValueChange": [Function],
              "handlePreciseXChange": [Function],
              "handlePreciseXValueChange": [Function],
              "handleRamXChange": [Function],
              "handleRamXValueChange": [Function],
              "handleSharpshooterXChange": [Function],
              "handleSharpshooterXValueChange": [Function],
              "handleSurgeConversionChange": [Function],
              "incrementAimTokenCount": [Function],
              "incrementBlackDice": [Function],
              "incrementRedDice": [Function],
              "incrementSurgeTokenCount": [Function],
              "incrementWhiteDice": [Function],
              "resetDiceCount": [Function],
              "resetTokenCounts": [Function],
            }
          }
          input={
            Object {
              "blackDice": 0,
              "blast": false,
              "criticalX": Object {
                "active": false,
                "value": 1,
              },
              "duelist": false,
              "highVelocity": false,
              "impactX": Object {
                "active": false,
                "value": 1,
              },
              "ionX": Object {
                "active": false,
                "value": 1,
              },
              "jediHunter": false,
              "lethalX": Object {
                "active": false,
                "value": 1,
              },
              "makashiMastery": false,
              "marksman": false,
              "pierceX": Object {
                "active": false,
                "value": 1,
              },
              "preciseX": Object {
                "active": false,
                "value": 1,
              },
              "ramX": Object {
                "active": false,
                "value": 1,
              },
              "redDice": 0,
              "sharpshooterX": Object {
                "active": false,
                "value": 1,
              },
              "surge": 1,
              "tokens": Object {
                "aim": 0,
                "surge": 0,
              },
              "whiteDice": 0,
            }
          }
          profileDialogId="attackProfileDialog"
          showSimpleView={true}
        />
      </div>
      <div
        className="col-md-5 col-lg-4"
      >
        <Defense
          eventHandlers={
            Object {
              "changeDie": [Function],
              "handleArmorChange": [Function],
              "handleArmorXChange": [Function],
              "handleArmorXValueChange": [Function],
              "handleBlockChange": [Function],
              "handleCoverChange": [Function],
              "handleDangerSenseXChange": [Function],
              "handleDangerSenseXValueChange": [Function],
              "handleDeflectChange": [Function],
              "handleDjemSoMasteryChange": [Function],
              "handleDuelistChange": [Function],
              "handleHasForceUpgradesChange": [Function],
              "handleImmuneBlastChange": [Function],
              "handleImmuneMeleePierceChange": [Function],
              "handleImmunePierceChange": [Function],
              "handleImperviousChange": [Function],
              "handleLowProfileChange": [Function],
              "handleOutmaneuverChange": [Function],
              "handleSoresuMasteryChange": [Function],
              "handleUncannyLuckXChange": [Function],
              "handleUncannyLuckXValueChange": [Function],
              "incrementDodgeTokenCount": [Function],
              "incrementObservationTokenCount": [Function],
              "incrementShieldTokenCount": [Function],
              "incrementSuppressionTokenCount": [Function],
              "incrementSurgeTokenCount": [Function],
              "resetTokenCount": [Function],
            }
          }
          input={
            Object {
              "armor": false,
              "armorX": Object {
                "active": false,
                "value": 1,
              },
              "block": false,
              "cover": "none",
              "dangerSenseX": Object {
                "active": false,
                "value": 1,
              },
              "deflect": false,
              "dieColor": "white",
              "djemSoMastery": false,
              "duelist": false,
              "hasForceUpgrades": false,
              "immuneBlast": false,
              "immuneMeleePierce": false,
              "immunePierce": false,
              "impervious": false,
              "lowProfile": false,
              "outmaneuver": false,
              "soresuMastery": false,
              "surge": true,
              "tokens": Object {
                "dodge": 0,
                "observation": 0,
                "shield": 0,
                "suppression": 0,
                "surge": 0,
              },
              "uncannyLuckX": Object {
                "active": false,
                "value": 1,
              },
            }
          }
          profileDialogId="defenseProfileDialog"
          showSimpleView={true}
        />
      </div>
      <div
        className="col-md-6 offset-md-3 col-lg-4 offset-lg-0 collapse"
      >
        <Combat
          eventHandlers={
            Object {
              "handleGuardianXChange": [Function],
              "handleGuardianXValueChange": [Function],
              "handleMeleeAttackChange": [Function],
            }
          }
          input={
            Object {
              "guardian": Object {
                "active": false,
                "value": 1,
              },
              "meleeAttack": false,
            }
          }
        />
      </div>
    </div>
    <div
      className="d-flex flex-wrap justify-content-center my-3"
    >
      <span
        className="mx-2 my-auto drop-down-label"
      >
        Rolls:
      </span>
      <select
        className="rounded-lg mr-4 px-2"
        onChange={[Function]}
        value={10000}
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="10000"
        >
          10,000
        </option>
      </select>
      <button
        className="btn btn-secondary border border-secondary rounded-lg mx-2 px-3 roll-button"
        onClick={[Function]}
      >
        Roll
      </button>
      <div
        className="d-flex justify-content-center my-auto custom-control custom-switch"
        key="showExpectedRangeToggle-abilitytoggle"
      >
        <input
          checked={true}
          className="custom-control-input my-auto collapse"
          id="showExpectedRangeToggle"
          key="showExpectedRangeToggle-toggle-input"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="custom-control-label drop-down-label mx-2 my-auto collapse"
          htmlFor="showExpectedRangeToggle"
          key="showExpectedRangeToggle-toggle-label"
        >
          Show expected range
        </label>
      </div>
    </div>
    <DiceResults
      results={
        Object {
          "firstAttack": Object {
            "attack": Object {
              "criticals": 0,
              "hits": 0,
              "misses": 0,
              "surges": 0,
            },
            "defense": Object {
              "blanks": 0,
              "blocks": 0,
              "forcedSaves": 0,
              "surges": 0,
              "wounds": 0,
            },
          },
          "summary": Object {
            "attackCount": 0,
            "attackSurge": Array [],
            "blocks": Array [],
            "critical": Array [],
            "defenseSurge": Array [],
            "forcedSaveStats": Object {
              "mean": 0,
              "median": 0,
              "stddev": 0,
            },
            "forcedSaves": Array [],
            "hit": Array [],
            "woundStats": Object {
              "mean": 0,
              "median": 0,
              "stddev": 0,
            },
            "wounds": Array [],
          },
        }
      }
      showExpectedRange={true}
      visibility={1}
    />
  </div>
</main>
`;

exports[`App matches the snapshot 1`] = `
<main
  role="main"
>
  <Header />
  <ProfileSelectorDialog
    applyProfile={[Function]}
    id="attackProfileDialog"
    upgradeAllowListName="attack"
  />
  <ProfileSelectorDialog
    applyProfile={[Function]}
    id="defenseProfileDialog"
    upgradeAllowListName="defense"
  />
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-8 offset-md-2 mt-3"
      >
        <div
          className="d-flex justify-content-center my-auto custom-control custom-switch"
          key="showSimplifiedViewToggle-abilitytoggle"
        >
          <input
            checked={false}
            className="custom-control-input my-auto"
            id="showSimplifiedViewToggle"
            key="showSimplifiedViewToggle-toggle-input"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="custom-control-label drop-down-label mx-2 my-auto"
            htmlFor="showSimplifiedViewToggle"
            key="showSimplifiedViewToggle-toggle-label"
          >
            Simplified View
          </label>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-5 offset-md-1 col-lg-4 offset-lg-0"
      >
        <Attack
          eventHandlers={
            Object {
              "handleBlastChange": [Function],
              "handleCriticalXChange": [Function],
              "handleCriticalXValueChange": [Function],
              "handleDuelistChange": [Function],
              "handleHighVelocityChange": [Function],
              "handleImpactXChange": [Function],
              "handleImpactXValueChange": [Function],
              "handleIonXChange": [Function],
              "handleIonXValueChange": [Function],
              "handleJediHunterChange": [Function],
              "handleLethalXChange": [Function],
              "handleLethalXValueChange": [Function],
              "handleMakashiMasteryChange": [Function],
              "handleMarksmanChange": [Function],
              "handlePierceXChange": [Function],
              "handlePierceXValueChange": [Function],
              "handlePreciseXChange": [Function],
              "handlePreciseXValueChange": [Function],
              "handleRamXChange": [Function],
              "handleRamXValueChange": [Function],
              "handleSharpshooterXChange": [Function],
              "handleSharpshooterXValueChange": [Function],
              "handleSurgeConversionChange": [Function],
              "incrementAimTokenCount": [Function],
              "incrementBlackDice": [Function],
              "incrementRedDice": [Function],
              "incrementSurgeTokenCount": [Function],
              "incrementWhiteDice": [Function],
              "resetDiceCount": [Function],
              "resetTokenCounts": [Function],
            }
          }
          input={
            Object {
              "blackDice": 0,
              "blast": false,
              "criticalX": Object {
                "active": false,
                "value": 1,
              },
              "duelist": false,
              "highVelocity": false,
              "impactX": Object {
                "active": false,
                "value": 1,
              },
              "ionX": Object {
                "active": false,
                "value": 1,
              },
              "jediHunter": false,
              "lethalX": Object {
                "active": false,
                "value": 1,
              },
              "makashiMastery": false,
              "marksman": false,
              "pierceX": Object {
                "active": false,
                "value": 1,
              },
              "preciseX": Object {
                "active": false,
                "value": 1,
              },
              "ramX": Object {
                "active": false,
                "value": 1,
              },
              "redDice": 0,
              "sharpshooterX": Object {
                "active": false,
                "value": 1,
              },
              "surge": 1,
              "tokens": Object {
                "aim": 0,
                "surge": 0,
              },
              "whiteDice": 0,
            }
          }
          profileDialogId="attackProfileDialog"
          showSimpleView={false}
        />
      </div>
      <div
        className="col-md-5 col-lg-4"
      >
        <Defense
          eventHandlers={
            Object {
              "changeDie": [Function],
              "handleArmorChange": [Function],
              "handleArmorXChange": [Function],
              "handleArmorXValueChange": [Function],
              "handleBlockChange": [Function],
              "handleCoverChange": [Function],
              "handleDangerSenseXChange": [Function],
              "handleDangerSenseXValueChange": [Function],
              "handleDeflectChange": [Function],
              "handleDjemSoMasteryChange": [Function],
              "handleDuelistChange": [Function],
              "handleHasForceUpgradesChange": [Function],
              "handleImmuneBlastChange": [Function],
              "handleImmuneMeleePierceChange": [Function],
              "handleImmunePierceChange": [Function],
              "handleImperviousChange": [Function],
              "handleLowProfileChange": [Function],
              "handleOutmaneuverChange": [Function],
              "handleSoresuMasteryChange": [Function],
              "handleUncannyLuckXChange": [Function],
              "handleUncannyLuckXValueChange": [Function],
              "incrementDodgeTokenCount": [Function],
              "incrementObservationTokenCount": [Function],
              "incrementShieldTokenCount": [Function],
              "incrementSuppressionTokenCount": [Function],
              "incrementSurgeTokenCount": [Function],
              "resetTokenCount": [Function],
            }
          }
          input={
            Object {
              "armor": false,
              "armorX": Object {
                "active": false,
                "value": 1,
              },
              "block": false,
              "cover": "none",
              "dangerSenseX": Object {
                "active": false,
                "value": 1,
              },
              "deflect": false,
              "dieColor": "white",
              "djemSoMastery": false,
              "duelist": false,
              "hasForceUpgrades": false,
              "immuneBlast": false,
              "immuneMeleePierce": false,
              "immunePierce": false,
              "impervious": false,
              "lowProfile": false,
              "outmaneuver": false,
              "soresuMastery": false,
              "surge": true,
              "tokens": Object {
                "dodge": 0,
                "observation": 0,
                "shield": 0,
                "suppression": 0,
                "surge": 0,
              },
              "uncannyLuckX": Object {
                "active": false,
                "value": 1,
              },
            }
          }
          profileDialogId="defenseProfileDialog"
          showSimpleView={false}
        />
      </div>
      <div
        className="col-md-6 offset-md-3 col-lg-4 offset-lg-0 collapse.show"
      >
        <Combat
          eventHandlers={
            Object {
              "handleGuardianXChange": [Function],
              "handleGuardianXValueChange": [Function],
              "handleMeleeAttackChange": [Function],
            }
          }
          input={
            Object {
              "guardian": Object {
                "active": false,
                "value": 1,
              },
              "meleeAttack": false,
            }
          }
        />
      </div>
    </div>
    <div
      className="d-flex flex-wrap justify-content-center my-3"
    >
      <span
        className="mx-2 my-auto drop-down-label"
      >
        Rolls:
      </span>
      <select
        className="rounded-lg mr-4 px-2"
        onChange={[Function]}
        value={10000}
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="10000"
        >
          10,000
        </option>
      </select>
      <button
        className="btn btn-secondary border border-secondary rounded-lg mx-2 px-3 roll-button"
        onClick={[Function]}
      >
        Roll
      </button>
      <div
        className="d-flex justify-content-center my-auto custom-control custom-switch"
        key="showExpectedRangeToggle-abilitytoggle"
      >
        <input
          checked={true}
          className="custom-control-input my-auto collapse"
          id="showExpectedRangeToggle"
          key="showExpectedRangeToggle-toggle-input"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="custom-control-label drop-down-label mx-2 my-auto collapse"
          htmlFor="showExpectedRangeToggle"
          key="showExpectedRangeToggle-toggle-label"
        >
          Show expected range
        </label>
      </div>
    </div>
    <DiceResults
      results={
        Object {
          "firstAttack": Object {
            "attack": Object {
              "criticals": 0,
              "hits": 0,
              "misses": 0,
              "surges": 0,
            },
            "defense": Object {
              "blanks": 0,
              "blocks": 0,
              "forcedSaves": 0,
              "surges": 0,
              "wounds": 0,
            },
          },
          "summary": Object {
            "attackCount": 0,
            "attackSurge": Array [],
            "blocks": Array [],
            "critical": Array [],
            "defenseSurge": Array [],
            "forcedSaveStats": Object {
              "mean": 0,
              "median": 0,
              "stddev": 0,
            },
            "forcedSaves": Array [],
            "hit": Array [],
            "woundStats": Object {
              "mean": 0,
              "median": 0,
              "stddev": 0,
            },
            "wounds": Array [],
          },
        }
      }
      showExpectedRange={true}
      visibility={1}
    />
  </div>
</main>
`;
